src/charm.py

```diff
@@ -459,95 +268,37 @@ def _requested_receivers(self) -> Tuple[ReceiverProtocol, ...]:
         requested_receivers = requested_protocols.intersection(set(self.tempo.receiver_ports))
         return tuple(requested_receivers)
 
-    def server_cert(self) -> str:
+    def server_ca_cert(self) -> str:
         """For charm tracing."""
-        self._update_server_cert()
-        return self.tempo.server_cert_path
-
-    def _update_server_cert(self) -> None:
-        """Server certificate for charm tracing tls, if tls is enabled."""
-        server_cert = Path(self.tempo.server_cert_path)
-        if self.tls_available:
-            if not server_cert.exists():
-                server_cert.parent.mkdir(parents=True, exist_ok=True)
-                if self.cert_handler.server_cert:
-                    server_cert.write_text(self.cert_handler.server_cert)
+        self._update_server_ca_cert()
+        return self.tempo.tls_ca_path
+
+    def _update_server_ca_cert(self) -> None:
+        """Server CA certificate for charm tracing tls, if tls is enabled."""
+        server_ca_cert = Path(self.tempo.tls_ca_path)
+        if self.coordinator.tls_available:
+            if not server_ca_cert.exists():
+                server_ca_cert.parent.mkdir(parents=True, exist_ok=True)
+                if self.coordinator.cert_handler.ca_cert:
+                    server_ca_cert.write_text(self.coordinator.cert_handler.ca_cert)
         else:  # tls unavailable: delete local cert
-            server_cert.unlink(missing_ok=True)
+            server_ca_cert.unlink(missing_ok=True)
 
     def tempo_otlp_http_endpoint(self) -> Optional[str]:
         """Endpoint at which the charm tracing information will be forwarded."""
         # the charm container and the tempo workload container have apparently the same
         # IP, so we can talk to tempo at localhost.
-        if self.tempo.is_ready:
+        if hasattr(self, "coordinator") and self.coordinator.tracing.is_ready():
```

Comment:
if the assumption in the above comment is correct, then self.coordinator would raise an `AttributeError` exception if it gets called before self.coordinator gets initialised.

Reply:
like above, I think it's fine to do `self.coordinator`

Reply:
(if it's not you would definitely catch this in unittests)

Reply:
(assuming we have at least one charm-tracing-enabled unittest)

Reply:
I think the issue is with `self._external_url` that gets passed to `Coordinator(external_url=self._external_url)`. That flow will eventually lead to `_internal_url` that will try to access `self.coordinator` that is not yet initialised. 
I think this might be an even broader problem? that `_external_url` will be passed to `MetricsEndpointProvider` and since `self.coordinator` will not be there yet, in the case of TLS, `MetricsEndpointProvider` will always get an http instead of https endpoint.

Reply:
ah. Well yes, that is definitely a problem. 
why does coordinator need external_url as an argument, given it wraps ingress as well? @lucabello

Reply:
it doesn't?

Reply:
I just realized I was writing on the wrong comment :) I'm referring to line 147
Shared coordinator doesn't wrap ingress. I think we can just remove the dependency on `self.coordinator.nginx.are_certificates_on_disk` and have a property in charm code that does that check for us.
```
    @property
    def are_certificates_on_disk(self) -> bool:
        """Return True if the certificates files are on disk."""
        nginx_container = self.unit.get_container("nginx")

        return (
            nginx_container.can_connect()
            and nginx_container.exists(CERT_PATH)
            and nginx_container.exists(KEY_PATH)
            and nginx_container.exists(CA_CERT_PATH)
        )

```
WDYT?

Reply:
The coordinator object does **not** wrap `IngressPerAppRequirer` specifically because apparently Tempo won't have that integration.

Correct me if I'm wrong, but shouldn't this same issue appear with anything that implements Ingress? All those charms have:

```python
@property
def external_url:
	return self.ingress.url or self.internal_url

def tempo_otlp_http_endpoint(self) -> Optional[str]:
	return self.external_url
```

Shouldn't this error for any charm related to ingress, considering "ingress" might not be yet initialized? This feels like something which is either a code ordering issue (which somehow doesn't happen in other places), or something that should be "fixed" in the tracing library itself if possible

Or maybe I'm just not clear on what's happening :)